p2p/status.go

```diff
@@ -0,0 +1,126 @@
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more detailg.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package p2p
+
+import (
+	"time"
+
+	"github.com/ChainSafe/gossamer/common"
+	log "github.com/ChainSafe/log15"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// SendStatusInterval is the time between sending status messages
+const SendStatusInterval = 5 * time.Minute
+
+// TODO: generate host status message
+var hostStatus = &StatusMessage{
+	ProtocolVersion:     0,
+	MinSupportedVersion: 0,
+	Roles:               0,
+	BestBlockNumber:     0,
+	BestBlockHash:       common.Hash{0x00},
+	GenesisHash:         common.Hash{0x00},
+	ChainStatus:         []byte{0},
+}
+
+// status submodule
+type status struct {
+	host        *host
+	confirmed   map[peer.ID]bool
+	lastMessage map[peer.ID]*StatusMessage
+}
+
+// newStatus creates a new status submodule
+func newStatus(host *host) (s *status, err error) {
+	s = &status{
+		host:        host,
+		confirmed:   make(map[peer.ID]bool),
+		lastMessage: make(map[peer.ID]*StatusMessage),
+	}
+	return s, err
+}
+
+// handleConn starts status processes upon connection
+func (s *status) handleConn(conn network.Conn) {
+
+	// starts sending status messages to connected peer
+	go s.sendMessages(conn.RemotePeer())
+
+}
+
+// sendMessages sends status messages to the connected peer
+func (status *status) sendMessages(peer peer.ID) {
+	for {
+		// TODO: use generated status message
+		msg := hostStatus
+
+		// send host status message to peer
+		err := status.host.send(peer, msg)
+		if err != nil {
+			log.Error(
+				"Failed to send status message to peer",
+				"peer", peer,
+				"err", err,
+			)
+		}
+
+		// wait between sending status messages
+		time.Sleep(SendStatusInterval)
+	}
+}
+
+// handleMessage checks if the status message is valid. If the status is valid,
+// handle confirmation and store latest status message, otherwise, ensure the
+// confirmation is false and close the connection with the peer.
+func (status *status) handleMessage(stream network.Stream, msg *StatusMessage) {
+	peer := stream.Conn().RemotePeer()
+
+	// check if valid status message
+	if status.validMessage(msg) {
+
+		// update peer status to true
+		status.confirmed[peer] = true
+
+		// update peer status message
+		status.lastMessage[peer] = msg
+
+		// TODO: unset per status after some time has passed
+
+	} else {
+
+		// ensure peer status is false
+		status.confirmed[peer] = false
+
+		// close connection with peer if status message is not valid
+		err := status.host.closePeer(peer)
```

Comment:
This could be replaced by `stream.Conn().Close()`

Reply:
It could. I think it was originally a method on my peer manager idea. I can fix.

Reply:
fixed